cip/1.accepted/CIP2017-04-20-query-combinators.adoc

```diff
@@ -0,0 +1,149 @@
+
+Multi-arm query combinators may determine the result signature of a top-level query.
+If any arm specifies record fields explicitly, the same set of record fields in exactly the same order is returned by the entire query.
+
+=== UNION
+
+`UNION` computes the logical set union between two sets of input records (i.e. any duplicates are discarded).
+
+`UNION MAX` computes the logical multiset union between two bags of input records (i.e. preserves the largest number of duplicates from either arm).
+
+`UNION ALL` computes the logical multiset union between two bags of input records (i.e. all duplicates from both arms are retained).
+
+=== INTERSECT
+
+`INTERSECT` computes the logical set intersection between two sets of input records (i.e. any duplicates are discarded).
+
+`INTERSECT ALL` computes the logical multiset intersection between two bags of input records (i.e. shared duplicates are retained).
+
+=== EXCEPT
+
+`EXCEPT` computes the logical set difference between two sets of input records (i.e. any duplicates are discarded).
+
+`EXCEPT ALL` computes the logical multiset difference between two bags of input records (i.e. excess duplicates on the left-hand side are retained).
+
+=== EXCLUSIVE UNION
+
+`EXCLUSIVE UNION` computes the exclusive logical set union between two sets of input records (i.e. any duplicates in the final outcome are discarded).
+
+`EXCLUSIVE UNION ALL` computes the exclusive logical multiset union between two bags of input records (i.e. the largest remaining excess multiplicity of each record in any argument bag is returned).
+
+=== OTHERWISE
+
+`OTHERWISE` computes the logical choice between two bags of input records.
+It evaluates to all records from the left-hand side argument provided the bag of input records is non-empty; otherwise it evaluates to all records from the right-hand side argument.
+
+=== CROSS
+
+`CROSS` computes the cartesian product between two bags of input records (i.e. preserves duplicates).
+
+In contrast to the other query combinators, the standard rules regarding returned record fields do not apply to `CROSS`.
+Instead, the set of returned record fields of both arms of a `CROSS` must be non-overlapping.
+The returned record fields of a `CROSS` operation consist of all the fields specified in the left arm (appearing in the order specified), followed by all the fields specified in the right arm (appearing in the order specified).
+
+=== THEN
+
+`THEN` computes query-level pipelining; i.e. it executes the right-hand side query for each input record from the left-hand side, and returns the flattened concatenation of all such records produced.
+
+The main feature of `THEN` is that it allows pipelining between nested subqueries.
+This is due to its syntactic status as a query combinator.
+
+In contrast to the other query combinators, the standard rules regarding returned record fields do not apply to `THEN`.
+Instead, the set of returned record fields of both arms of `THEN` must be non-overlapping.
```

Comment:
Why is this required? Isn't this query perfectly reasonable:

```
RETURN 1 AS one, 2 AS two
THEN
RETURN one + 5 AS one, one + two AS three
```

And the full statement would return the fields `one` and `three`?